### PR TITLE
power: blizzard's atlas for insanity, maelstorm, lunar power bars

### PR DESCRIPTION
**Pull request by bbeny123**
Also updates data from FrameXML/Constants.lua required to proper working of this feature

### CODE REVIEW

**Comment by bbeny123** [comment]
changes

**Comment by ls-** [comment]
Hi!

There's one big issue w/ this PR, it's quite dirty. You should base it on [master's version of power](https://github.com/haste/oUF/blob/master/elements/power.lua) without code from other patches/PRs. I can see code from my power fix PR, p3lim's arena prep frames, also try not to include things that aren't related to what this PR is supposed to do, like do not clean trailing whitespaces, etc.

**Comment by ls-** [comment]
K. one issue [here ](https://github.com/bbeny123/oUF/blob/15597b29e31ecafb19f7c736b612b96f8fd8ae66/elements/power.lua#L221-L230), if there a status bar texture set in layout, it shouldn't happen at all.

**Comment by bbeny123** [comment]
Hi

> There's one big issue w/ this PR, it's quite dirty. You should base it on master's version of power without code from other patches/PRs 

Sorry for that, initially i forked your rep not haste's. I already changed this, and now this pull request added only legion style power bars :)

**Comment by p3lim** [comment]
Please keep your PR clean, don't include changes that are irrelevant to the feature you're promoting, and once you're done committing, please rebase the commit(s) to as few as needed, formatting the commit message as such: "elementname: changes".

As far as the feature goes, it's a neat idea, but how it's enabled needs to be changed to be optional, currently it's considered opt-out.

**Comment by bbeny123** [comment]
Now should be ok, its fully optional and i merged all commits to one :)

**Comment by Blazeflack** [comment]
The commit still removes beta checks. These changes are irrelevant to this feature and should not be included.

**Comment by p3lim** [comment]
It should also be tied to the `colorPower` option imo.

**Comment by bbeny123** [comment]
> The commit still removes beta checks. These changes are irrelevant to this feature and should not be included.

fixed

> It should also be tied to the colorPower option imo.

added

**Comment by Blazeflack** [comment]
>fixed

Not quite yet. You're still making changes that are unrelated to this feature.

See this:
http://i.imgur.com/6ohNcc7.png

**Comment by bbeny123** [comment]
This changes are related, because this feature using a data from FrameXML/Constants.lua. 
i.e. this feature wouldn't work for chi, if chi would have atlas (without this changes)

**Comment by p3lim** [comment]
- Chi has always been been on index 12, even back when it was called "Light Force" (in MoP alpha they were toying with the idea of having light and dark chi, a yin-yan theme as their resource).
- Combopoints has always been on index 4.
- Arcane charges didn't exist in WoD

[source](https://www.townlong-yak.com/framexml/beta/Constants.lua/diff#108)

Regardless, even if it _was_ wrong, it would be a bug and required a different PR.

**Comment by bbeny123** [comment]
> Regardless, even if it was wrong, it would be a bug and required a different PR.

Ok, i revetred this change. I will make a different PR for it.

**Comment by ls-** [comment]
@p3lim, there's some inconsistency on Blizz side, yet again, [here](https://www.townlong-yak.com/framexml/beta/UnitFrame.lua#30), you can see that Chi sits at 4 T_T

**Comment by p3lim** [comment]
@ls- The reason we follow the ones in Constants.lua is because we're using those constants.

**Comment by ls-** [comment]
@p3lim, yeah, Ikr T_T I just wanted to show, that there's inconsistency and person probably was confused by it.

**Comment by gagou** [comment]
I think that we should have an option to use the power atlas for the bg if it exists.


Currently bg color is not set correctly when using an atlas as r, g, b are not given values.

Instead of doing:
`power:SetStatusBarColor(1, 1, 1)`

I would do:
`r, g, b = 1, 1, 1`

File: elements/power.lua [hunk 1]
```diff
@@ -258,7 +276,10 @@ local Enable = function(self, unit)
 		self:RegisterEvent('UNIT_FACTION', Path)
 
 		if(power:IsObjectType'StatusBar' and not power:GetStatusBarTexture()) then
-			power:SetStatusBarTexture[[Interface\TargetingFrame\UI-StatusBar]]
+			if(power.colorPower and power.statusbar and power.atlasEnable) then
+				statusbar = power.statusbar
```
Comment (p3lim) [inline elements/power.lua]: The layout should not be required to add extra information for internal logic, a better way of handling this is to set the original statusbar variable _after_ the fallback code, and just use normal methods of getting it.

Basically, after [this line](https://github.com/haste/oUF/blob/441ee4be0d1140f7861e416609760cf7003a36be/elements/power.lua#L262) you could do `statusbar = power:GetStatusBarTexture()` (but you also have to consider the posibility that the object might not be a statusbar, requiring a teensy bit more logic).

Also, different statusbars might have different textures, storing them globally for the entire element is not a good idea, you should store it on the element itself (e.g. `power._originalTexture`).

File: elements/power.lua [hunk 2]
```diff
@@ -208,9 +212,23 @@ local Update = function(self, event, unit)
         local adjust = 0 - (min or 0)
 		r, g, b = self.ColorGradient(cur + adjust, max + adjust, unpack(power.smoothGradient or self.colors.smooth))
 	end
-
+	
+	if(power.statusbar) then
```
Comment (p3lim) [inline elements/power.lua]: What is the purpose of this? It's basically a duplicate of line 280.

File: elements/power.lua [hunk 3]
```diff
@@ -208,9 +212,23 @@ local Update = function(self, event, unit)
...
+			else
+				power:SetStatusBarTexture(statusbar)
+				r, g, b = t[1], t[2], t[3]
+			end
+		else
+			r, g, b = t[1], t[2], t[3]
+		end
 	end
 
 	if(b) then
```
Comment (p3lim) [inline elements/power.lua]: The variables `r`, `g` and `b` can still exist here, even if atlas is used, and will result in a discolored bar.

File: elements/power.lua [hunk 4]
```diff
@@ -208,9 +212,23 @@ local Update = function(self, event, unit)
         local adjust = 0 - (min or 0)
 		r, g, b = self.ColorGradient(cur + adjust, max + adjust, unpack(power.smoothGradient or self.colors.smooth))
 	end
-
+	
+	if(power.statusbar) then
+		statusbar = power.statusbar
+	end
+
```
Comment (p3lim) [inline elements/power.lua]: What is the purpose of this block?

File: elements/power.lua [hunk 5]
```diff
@@ -208,9 +212,23 @@ local Update = function(self, event, unit)
         local adjust = 0 - (min or 0)
 		r, g, b = self.ColorGradient(cur + adjust, max + adjust, unpack(power.smoothGradient or self.colors.smooth))
 	end
-
+	
+	if(power.statusbar) then
+		statusbar = power.statusbar
+	end
```
Comment (p3lim) [inline elements/power.lua]: What is the purpose of this block?

File: elements/power.lua (hunk 3 above)
Comment (p3lim) [inline elements/power.lua]: The variables r, g and b can still exist even if atlas is used, and will result in a discolored bar (although the situation might be rare, it's best to have a failsafe).

File: elements/power.lua (hunk 1 above)
Comment (p3lim) [inline elements/power.lua]: The layout should not be required to add extra information for internal logic, a better way of handling this is to set the original statusbar variable after the fallback code, and just use normal methods of getting it.

Basically, after [this line](https://github.com/oUF-wow/oUF/blob/441ee4be0d1140f7861e416609760cf7003a36be/elements/power.lua#L262) you could do statusbar = power:GetStatusBarTexture() (but you also have to consider the posibility that the object might not be a statusbar, requiring a teensy bit more logic).

Also, different statusbars might have different textures, storing them globally for the entire element is not a good idea, you should store it on the element itself (e.g. power._originalTexture).

**Comment by ls-** [comment]
Closing this one in favour of #304.

@bbeny123, thanks for the idea! 👍

**Pull request closed without merging**